### ISSUE TITLE
Performance optimization

### ISSUE THREAD

**Issue by JamesBremner** [JamesBremner/so76550312#2]
Required to optimize the performance of create_edges_and_nodes_from_profiles()

**Comment by JamesBremner** [comment]
Detailed timing profile 
```

raven::set::cRunWatch code timing profile
Calls           Mean (secs)     Total           Scope
       1        0.220081        0.220081        create_edges_and_nodes_from_profiles
   10000        2.05228e-05     0.205228        loop over profiles
   10000        7.79368e-06     0.0779368       rest of profile loop
   10000        5.6973e-06      0.056973        populate edges
  100000        1.38977e-07     0.0138977       populate nodes
  100000        1.11064e-07     0.0111064       loop over experiences
   10000        1.00089e-06     0.0100089       reserve
   10000        5.643e-08       0.0005643       populate company map
```

**Comment by JamesBremner** [comment]
@IshanHegde I would like to discuss this issue here, rather than on stackoverflow.  I find it much easier to use - hope you agree.

**Comment by JamesBremner** [comment]
Notice this code:

https://github.com/JamesBremner/so76550312/blob/4af272f5613894dae0ea53cf1ea99009945ed41a/src/profiles.cpp#L44-L51

Each time through this loop the node_temp vector is constructed and then deleted.  Just doing the memory reservation takes a total of 10 msecs, and the deletion will take at least is long ( I do not have a measurement for deletion )

If node_temp is constructed outside the loop and simply cleared then, as expected, about 20 msecs are saved

```
raven::set::cRunWatch code timing profile
Calls           Mean (secs)     Total           Scope
       1        0.201701        0.201701        create_edges_and_nodes_from_profiles
   10000        1.88924e-05     0.188924        loop over profiles
   10000        7.13995e-06     0.0713995       rest of profile loop
   10000        4.72151e-06     0.0472151       populate edges
  100000        1.69142e-07     0.0169142       populate nodes
  100000        1.08873e-07     0.0108873       loop over experiences
   10000        3.619e-08       0.0003619       populate company map
```

**Comment by JamesBremner** [comment]
Notice this code

https://github.com/JamesBremner/so76550312/blob/ef65a3b4487553145bfcca1fc087cdfa7f5f7d70/src/definitions.h#L182-L185

Every edge stores a complete copy of both the source and target node.  This is bad because:

1. It wastes an enormous amount of memory
2. It reduces performance of edge construction due to numerous calls to node copy constructor
3. Tends to generate bugs because of the data duplication requires careful updating of node attributes

I have replaced the node copies with indices to the nodes in the `node` vector.  I have also changed the edge constructor to use constant references to the nodes - saving another two calls to the node copy constructor.

This saves 50 msecs

```
raven::set::cRunWatch code timing profile
Calls           Mean (secs)     Total           Scope
       1        0.150164        0.150164        create_edges_and_nodes_from_profiles
   10000        1.38209e-05     0.138209        loop over profiles
   10000        3.67751e-06     0.0367751       rest of profile loop
   10000        2.95442e-06     0.0295442       populate nodes
  100000        1.04422e-07     0.0104422       loop over experiences
   10000        5.6704e-07      0.0056704       populate edges
   10000        3.354e-08       0.0003354       populate company map
```

**Comment by ghost** [comment]
Thanks a Ton for the detailed analysis. I also got performence improvements using member initializer lists for the data objects. Using Node indexes in Edges is a brilliant idea.

**Comment by JamesBremner** [comment]
Notice this code

https://github.com/JamesBremner/so76550312/blob/2146056f8f7f421f88b4914e8d897e78c1aaf4ea/src/profiles.cpp#L110-L111

You check if the profile is used ( 1 or more experience with a defined location ) after you construct the profile nodes.  If it is not used then you discard all the nodes you have just constructed!  It makes more sense to check if the profile is used first, and only construct nodes if they are going to be used.  I have implemented this change in commit https://github.com/JamesBremner/so76550312/commit/433b8b58bebe9594fffac75d45a2559c4c8e85aa

**Comment by JamesBremner** [comment]
Notice this code

https://github.com/JamesBremner/so76550312/blob/433b8b58bebe9594fffac75d45a2559c4c8e85aa/src/main.cpp#L33-L34

The entire purpose of all this code is to create a graph of nodes and edges, but you do not have a graph class to do the work.

It would be more convenient and faster to have something like this:

    // Create nodes and edges from profiles and companies
    cProjectGraph g2( vProfile, vCompany );

which invokes the graph class constructor to look after everything efficiently.

Implementing this requires many changes throughout the code, so I decided to fix up everything all at once, rather than tackling each small problem one by one.

Here is the code for the graph class constructor

https://github.com/JamesBremner/so76550312/blob/f70feca090e55d9cef8db938f7a1aee3e4ea3272/src/cProjectGraph.cpp#L3-L43

I think this is much easier to understand and maintain.  In any case, it runs about 20 times faster!

```
100000 nodes and 90000 edges constructed
raven::set::cRunWatch code timing profile
Calls           Mean (secs)     Total           Scope
       1        0.0053671       0.0053671       cProjectGraph constructor
```
I have committed the code for this to the "james" branch.
